tests/unit/contentproviders/test_hydroshare.py

```diff
@@ -0,0 +1,212 @@
+import json
+import os
+import pytest
+
+from contextlib import contextmanager
+from io import BytesIO
+from tempfile import TemporaryDirectory, NamedTemporaryFile
+from unittest.mock import patch
+from urllib.request import urlopen, Request, urlretrieve
+from zipfile import ZipFile
+
+from repo2docker.contentproviders import Hydroshare
+from repo2docker.contentproviders.base import ContentProviderException
+
+
+def test_content_id():
+    with patch.object(Hydroshare, "urlopen") as fake_urlopen:
+        fake_urlopen.return_value.url = (
+            "https://www.hydroshare.org/resource/b8f6eae9d89241cf8b5904033460af61"
+        )
+
+        def read():
+            return '{"dates": [{"type": "modified", "start_date": "2019-09-25T16:09:17.006152Z"}]}'
+
+        fake_urlopen.return_value.read = read
+        hydro = Hydroshare()
+
+        hydro.detect("10.4211/hs.b8f6eae9d89241cf8b5904033460af61")
+        assert hydro.content_id == "b8f6eae9d89241cf8b5904033460af61.v1569427757"
+
+
+test_hosts = [
+    (
+        [
+            "https://www.hydroshare.org/resource/b8f6eae9d89241cf8b5904033460af61",
+            "10.4211/hs.b8f6eae9d89241cf8b5904033460af61",
+            "https://doi.org/10.4211/hs.b8f6eae9d89241cf8b5904033460af61",
+        ],
+        {
+            "host": {
+                "hostname": [
+                    "https://www.hydroshare.org/resource/",
+                    "http://www.hydroshare.org/resource/",
+                ],
+                "django_irods": "https://www.hydroshare.org/django_irods/download/bags/",
+                "version": "https://www.hydroshare.org/hsapi/resource/{}/scimeta/elements",
+            },
+            "resource": "b8f6eae9d89241cf8b5904033460af61",
+            "version": "1569427757",
+        },
+    )
+]
+
+
+@pytest.mark.parametrize("test_input,expected", test_hosts)
+def test_detect_hydroshare(test_input, expected):
+    with patch.object(Hydroshare, "urlopen") as fake_urlopen:
+        fake_urlopen.return_value.url = test_input[0]
+
+        def read():
+            return '{"dates": [{"type": "modified", "start_date": "2019-09-25T16:09:17.006152Z"}]}'
+
+        fake_urlopen.return_value.read = read
+        # valid Hydroshare DOIs trigger this content provider
+        assert Hydroshare().detect(test_input[0]) == expected
+        assert Hydroshare().detect(test_input[1]) == expected
+        assert Hydroshare().detect(test_input[2]) == expected
+        # only two of the three calls above have to resolve a DOI
+        assert fake_urlopen.call_count == 5
```

Comment:
Why do we expect five calls here? From the comment above I was guessing that this should be two.

Reply:
It's 5 because each of the DOI calls call urlopen twice (once to fetch the version and another to resolve the DOI to the Hydroshare url).  The HydroShare url only makes the call to fetch version.  I can see how this is unclear, I'll update comments to clarify.